### PR TITLE
tests/robustness: fix out of index panic in model replay

### PR DESCRIPTION
**Pull request by MadhavJivrajani**
Encountered this while addressing comments in https://github.com/etcd-io/etcd/pull/17352

Panic encountered:
```
--- FAIL: TestRobustnessExploratory (5.57s)
    --- FAIL: TestRobustnessExploratory/Kubernetes/LowTraffic/ClusterOfSize1 (5.55s)
panic: runtime error: index out of range [184] with length 184 [recovered]
	panic: runtime error: index out of range [184] with length 184

goroutine 162 [running]:
testing.tRunner.func1.2({0x1a829c0, 0xc0001b5320})
	/usr/local/go/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1548 +0x397
panic({0x1a829c0?, 0xc0001b5320?})
	/usr/local/go/src/runtime/panic.go:914 +0x21f
go.etcd.io/etcd/tests/v3/robustness/model.(*EtcdReplay).next(0xc0000c4d30)
	/Users/mjivrajani/gocode/src/github.com/MadhavJivrajani/etcd/tests/robustness/model/replay.go:76 +0x585
go.etcd.io/etcd/tests/v3/robustness/model.(*EtcdReplay).StateForRevision(0xc0002aad30, 0xb9)
	/Users/mjivrajani/gocode/src/github.com/MadhavJivrajani/etcd/tests/robustness/model/replay.go:52 +0x1c5
go.etcd.io/etcd/tests/v3/robustness/validate.validatePrevKV(0xc000500820, {0x2, {0x23758c0, 0x0, 0x0}, {0xc00049c340, 0x1, 0x1}}, {0xc00043e000, 0xb8, ...})
	/Users/mjivrajani/gocode/src/github.com/MadhavJivrajani/etcd/tests/robustness/validate/watch.go:158 +0x2cf
go.etcd.io/etcd/tests/v3/robustness/validate.validateWatch(0xa?, 0x1f3?, {0x19?}, {0xc0005fa380, 0xa, 0xb27b95a0?}, {0xc00043e000, 0xb8, 0x110}, 0x1)
	/Users/mjivrajani/gocode/src/github.com/MadhavJivrajani/etcd/tests/robustness/validate/watch.go:39 +0x2fd
go.etcd.io/etcd/tests/v3/robustness/validate.ValidateAndReturnVisualize(0xc000500820, 0xc0005fa380?, {0x0?}, {0xc0005fa380, 0xa, 0x10}, 0x0?)
	/Users/mjivrajani/gocode/src/github.com/MadhavJivrajani/etcd/tests/robustness/validate/validate.go:45 +0x205
go.etcd.io/etcd/tests/v3/robustness.testRobustness({_, _}, _, _, {{0xc0000482d0, 0x24}, {0x1c6d200, 0x2328e40}, {{{0x1afe9b2, 0x7}, ...}, ...}, ...})
	/Users/mjivrajani/gocode/src/github.com/MadhavJivrajani/etcd/tests/robustness/main_test.go:97 +0x36b
go.etcd.io/etcd/tests/v3/robustness.TestRobustnessExploratory.func1(0xc000500820?)
	/Users/mjivrajani/gocode/src/github.com/MadhavJivrajani/etcd/tests/robustness/main_test.go:49 +0xb2
testing.tRunner(0xc000500820, 0xc0003ac610)
	/usr/local/go/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 128
	/usr/local/go/src/testing/testing.go:1648 +0x3ad

```

/assign @serathius

### CODE REVIEW

File: tests/robustness/model/replay.go [hunk 1]
```diff
@@ -73,7 +73,7 @@ func (r *EtcdReplay) next() (request EtcdRequest, revision int64, index int) {
 	revision = r.eventHistory[r.eventHistoryIndex].Revision
 	index = r.eventHistoryIndex
 	operations := []EtcdOperation{}
-	for r.eventHistory[index].Revision == revision {
+	for index < len(r.eventHistory) && r.eventHistory[index].Revision == revision {
```
Comment (serathius) [inline tests/robustness/model/replay.go:76]:
```suggestion
	if index == len(r.eventHistory) {
		panic("Cannot progress further, exhausted the whole event history")
	}
	for index < len(r.eventHistory) && r.eventHistory[index].Revision == revision {
```

Comment (serathius) [inline tests/robustness/model/replay.go:76]:
```suggestion
	if index == len(r.eventHistory) {
		panic("Cannot progress further, exhausted whole event history")
	}
	for index < len(r.eventHistory) && r.eventHistory[index].Revision == revision {
```

Comment (serathius) [reply tests/robustness/model/replay.go:76]: Nvm, badly handled edge case.

**Comment by MadhavJivrajani** [comment]
Ah, that makes sense - thanks! Let me fix that.

File: tests/robustness/model/replay.go (hunk 1 above)
Comment (serathius) [inline tests/robustness/model/replay.go:76]:
```suggestion
	if index > len(r.eventHistory) {
		panic("Cannot progress further, exhausted whole event history")
	}
	for index < len(r.eventHistory) && r.eventHistory[index].Revision == revision {
```

Comment (MadhavJivrajani) [reply tests/robustness/model/replay.go:76]: Coming into `next` I don't think we will ever hit that because of the loop condition in `StateForRevision`, when `index++` happens, we can go out of range.

Comment (serathius) [reply tests/robustness/model/replay.go:76]: you are right. The `StateForRevision` code will correctly prevent overflowing index.

The only thing we need to fix is `next` function called on the last revision that incorrectly panic as it try to check if revision still continues under the next index.

Your fix is correct and doesn't require any changes. Thanks!

One request for next PR would be to provide more context to make review easier.

**Review by serathius — APPROVED** [review]
